accepted/0000-nohoist.md

```diff
@@ -0,0 +1,299 @@
+1. tools like grunt assumes directory structure.
+    * this could require nohoist both shallow and deep dependency trees
+1. some projects may prefer no-hoisting for whatever reason ;-)
+
+## nohoist address a common issue in hoisted monorepo project
+
+`nohoist` is not a new concept, [lerna](https://github.com/lerna/lerna#--nohoist-glob) has provided a similar feature, an indicator that this is a common issue many monorepo projects have to deal with.
+
+# Detailed design
+
+## principle
+- hoist as much as we can
+- nohoist is an exception/workaround, therefore favor 
+    - explicitness over convenience
+    - small scale over large scale
+
+These principles lead us to the "**shallow nohoist**" implementation: _nohoist only apply to the package explicitly specified; all of its dependency, unless also on the nohoist list, will be hoisted by default._
+
+## divide the problem 
+
+nohoist is meant to address the hoist side-effect, thus break down by scale, from hoisting all to none:
+```
+hoist-all(1) --> nohoist-direct-dep(2) --> nohoist-deep-dep(3) --> hoist-none(4)
+```
+1. hoist-all: hoisting everything, the current behavior of yarn workspaces.
+2. nohoist-direct-dep: opt-out hoisting for direct dependency only, use-case #1
+3. nohoist-deep-dep: opt-out hoisting even for dependencies' dependencies... Maybe use-case #2
+4. hoist-none: a.k.a nohoist all, opt-out hoisting all together: use-case #3.
+
+yarn currently support (1), this RFC proposed solution to handle (2), (3), (4).
+
+## configuration
+We need a place to specify "nohoist" packages for the given workspace(s). We can expand the existing [workspaces](https://yarnpkg.com/en/docs/workspaces) config:
+```
+export type WorkspacesConfig = {
+  packages?: Array<string>,
+  nohoist?: Array<string>,
+};
+
+export type Manifest = {
+    ...
+    workspaces?: Array<string> | WorkspacesConfig
+}
+```
+note: the config will be backward compatible.
+
+### nohoist list inheritance
+nohoist config is inherited, i.e. if nohoist is specified in the root package.json, all workspaces will inherit the nohoist list, likewise all the dependencies of the workspace will inherit the nohoist list. This concept is simple and useful not only in reasoning nohoist from root to workspaces, it can be extended downwards to allow packages provide better monorepo support as well as easily handling deep-dependency as direct-dependency. 
+
+### nohoist list naming
+lerna uses glob pattern to specify nohoist package, however based on the design principle, I argue it is probably better not to since nohoist should be "rare", therefore probably doesn't justify for extra matching complexity and risk. 
+
+However, it is not feasible to ask developers mark off every package if they just want to opt-out hoisting all together, therefore a simple keyword like "\_all\_" should be introduced to handle use-case-3 easily. For example the following config will basically turn off nohoist for the whole monorepo project:

```

Comment:
I am not a huge fan of a new keyword.
Would it really be simpler than a glob?

Considering that we use globs already I think it would be expected by final users that nohoist fields is a glob too

Reply:
@bestander I see your point, yes we can indeed use glob pattern to make it consistent with the packages patterns. 

BTW, I loved the "why" feature of yarn, however, while using it to diagnose nohoist, there are a few bugs surfaced, mainly the current implementation assumed a given package only appeared once, which is not always true: 1. when version conflict 2. when it is marked nohoist. I will address them also in this PR so people can easily find out why a package appeared in multiple places.

(There is another why.js bug that it didn't look at peerDependency, but I will not include it in this PR, as it is independent of nohoist, trying to keep the scope tight.)